core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -310,6 +312,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
```

Comment:
another thing that hit me late last night is that this only holds for the cluster service. The transport service needs to be started to have the published address, which is needed to create the local node